File: scripts/generate-registry.ts
#!/usr/bin/env tsx

import { readFile, writeFile, mkdir } from 'fs/promises';
import { glob } from 'glob';
import matter from 'gray-matter';
import path from 'path';
import { fileURLToPath } from 'url';

// ============================================================================
// Type Definitions
// ============================================================================

interface AgentFrontmatter {
  name: string;
  description: string;
  tools?: string[];
  model?: string;
  model_rationale?: string;
  [key: string]: unknown;
}

interface AgentEntry {
  name: string;
  tier: string;
  category: string;
  complexity: string;
  model: string;
  description: string;
  capabilities: string[];
  tags: string[];
  file: string;
}

interface RegistryStats {
  total: number;
  by_tier: Record<string, number>;
  by_category: Record<string, number>;
  by_complexity: Record<string, number>;
  by_model: Record<string, number>;
}

interface AgentRegistry {
  generated: string;
  version: string;
  agents: AgentEntry[];
  stats: RegistryStats;
}

// ============================================================================
// Configuration
// ============================================================================

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const REPO_ROOT = path.join(__dirname, '..');
const AGENTS_DIR = path.join(REPO_ROOT, 'agents');
const OUTPUT_PATH = path.join(REPO_ROOT, 'configs', 'agent-registry.json');

// ============================================================================
// Utility Functions
// ============================================================================

/**
 * Extracts tier from file path (e.g., "01-foundation" from "agents/01-foundation/api-platform-engineer.md")
 */
function extractTier(filePath: string): string {
  const parts = filePath.split(path.sep);
  const tierPart = parts.find(part => /^\d{2}-/.test(part));
  return tierPart || 'uncategorized';
}

/**
 * Determines category based on tier
 */
function determineCategory(tier: string): string {
  const tierMap: Record<string, string> = {
    '00-meta': 'orchestration',
    '01-foundation': 'foundation',
    '02-development': 'development',
    '03-specialists': 'specialist',
    '04-experts': 'expert',
    '05-domains': 'domain',
    '06-integration': 'integration',
    '07-quality': 'quality',
    '08-automation': 'automation',
    '08-finance': 'finance',
    '09-enterprise': 'enterprise',
  };
  return tierMap[tier] || 'uncategorized';
}

/**
 * Determines complexity based on description length and tool count
 */
function determineComplexity(description: string, tools: string[]): string {
  const descLength = description.length;
  const toolCount = tools.length;

  if (descLength > 300 || toolCount >= 6) return 'complex';
  if (descLength > 150 || toolCount >= 4) return 'moderate';
  return 'simple';
}

/**
 * Extracts capabilities from description (capitalized keywords)
 */
function extractCapabilities(description: string): string[] {
  // Extract key technical terms and patterns
  const capabilities: string[] = [];
  const keywords = [
    'REST', 'GraphQL', 'gRPC', 'API', 'OpenAPI', 'OAuth', 'JWT',
    'React', 'Vue', 'Angular', 'TypeScript', 'JavaScript', 'Python', 'Go', 'Java',
    'AWS', 'Azure', 'GCP', 'Kubernetes', 'Docker', 'Terraform',
    'PostgreSQL', 'MongoDB', 'Redis', 'DynamoDB',
    'CI/CD', 'DevOps', 'MLOps', 'SRE', 'Security',
    'Machine Learning', 'Deep Learning', 'NLP', 'Computer Vision',
    'Testing', 'Performance', 'Observability', 'Monitoring',
    'Frontend', 'Backend', 'Full Stack', 'Mobile',
    'Microservices', 'Event-Driven', 'CQRS', 'DDD',
  ];

  for (const keyword of keywords) {
    const regex = new RegExp(`\\b${keyword}\\b`, 'i');
    if (regex.test(description)) {
      capabilities.push(keyword);
    }
  }

  return [...new Set(capabilities)].slice(0, 10); // Limit to 10 unique capabilities
}

/**
 * Extracts tags from description (lowercased keywords)
 */
function extractTags(description: string, name: string): string[] {
  const tags = new Set<string>();

  // Add name-based tags
  const nameParts = name.split('-');
  nameParts.forEach(part => {
    if (part.length > 3) tags.add(part);
  });

  // Extract tags from description
  const tagKeywords = [
    'api', 'rest', 'graphql', 'design', 'architecture', 'security',
    'testing', 'performance', 'optimization', 'monitoring', 'observability',
    'cloud', 'aws', 'kubernetes', 'docker', 'devops', 'ci/cd',
    'frontend', 'backend', 'mobile', 'react', 'typescript', 'python',
    'database', 'sql', 'nosql', 'machine learning', 'ml', 'ai',
    'documentation', 'research', 'review', 'debugging',
  ];

  tagKeywords.forEach(keyword => {
    const regex = new RegExp(`\\b${keyword}\\b`, 'i');
    if (regex.test(description)) {
      tags.add(keyword.toLowerCase());
    }
  });

  return Array.from(tags).slice(0, 15); // Limit to 15 tags
}

/**
 * Normalizes model name to standard format
 */
function normalizeModel(model?: string): string {
  if (!model) return 'sonnet';

  const modelLower = model.toLowerCase();
  if (modelLower.includes('opus')) return 'opus';
  if (modelLower.includes('sonnet')) return 'sonnet';
  if (modelLower.includes('haiku')) return 'haiku';

  return 'sonnet'; // Default
}

// ============================================================================
// Core Functions
// ============================================================================

/**
 * Parses a single agent markdown file and extracts metadata
 */
async function parseAgentFile(filePath: string): Promise<AgentEntry | null> {
  try {
    const content = await readFile(filePath, 'utf-8');
    const { data } = matter(content);

    // Validate required fields
    if (!data.name || !data.description) {
      console.warn(`⚠ Skipping ${filePath}: missing required frontmatter fields`);
      return null;
    }

    const frontmatter = data as AgentFrontmatter;
    const tier = extractTier(filePath);
    const category = determineCategory(tier);
    const tools = Array.isArray(frontmatter.tools)
      ? frontmatter.tools
      : typeof frontmatter.tools === 'string'
        ? frontmatter.tools.split(/[,\s]+/).filter(Boolean)
        : [];

    const complexity = determineComplexity(frontmatter.description, tools);
    const model = normalizeModel(frontmatter.model);
    const capabilities = extractCapabilities(frontmatter.description);
    const tags = extractTags(frontmatter.description, frontmatter.name);
    const relativeFile = path.relative(REPO_ROOT, filePath);

    return {
      name: frontmatter.name,
      tier,
      category,
      complexity,
      model,
      description: frontmatter.description.trim(),
      capabilities,
      tags,
      file: relativeFile,
    };
  } catch (error) {
    console.error(`✗ Error parsing ${filePath}:`, error);
    return null;
  }
}

/**
 * Scans all agent files and builds registry entries
 */
async function scanAgents(): Promise<AgentEntry[]> {
  const pattern = path.join(AGENTS_DIR, '**', '*.md');
  const files = await glob(pattern, {
    ignore: ['**/README.md', '**/AGENT_CHECKLIST.md', '**/TESTING.md'],
  });

  console.log(`Found ${files.length} agent files to process...`);

  const agents: AgentEntry[] = [];
  for (const file of files) {
    const agent = await parseAgentFile(file);
    if (agent) {
      agents.push(agent);
    }
  }

  // Sort by tier then name
  return agents.sort((a, b) => {
    const tierCompare = a.tier.localeCompare(b.tier);
    return tierCompare !== 0 ? tierCompare : a.name.localeCompare(b.name);
  });
}

/**
 * Calculates statistics for the registry
 */
function calculateStats(agents: AgentEntry[]): RegistryStats {
  const stats: RegistryStats = {
    total: agents.length,
    by_tier: {},
    by_category: {},
    by_complexity: {},
    by_model: {},
  };

  for (const agent of agents) {
    // Count by tier
    stats.by_tier[agent.tier] = (stats.by_tier[agent.tier] || 0) + 1;

    // Count by category
    stats.by_category[agent.category] = (stats.by_category[agent.category] || 0) + 1;

    // Count by complexity
    stats.by_complexity[agent.complexity] = (stats.by_complexity[agent.complexity] || 0) + 1;

    // Count by model
    stats.by_model[agent.model] = (stats.by_model[agent.model] || 0) + 1;
  }

  return stats;
}

/**
 * Generates the complete registry JSON
 */
async function generateRegistry(): Promise<void> {
  try {
    console.log('🚀 Generating agent registry...\n');

    const agents = await scanAgents();

    if (agents.length === 0) {
      throw new Error('No valid agent files found');
    }

    const stats = calculateStats(agents);

    const registry: AgentRegistry = {
      generated: new Date().toISOString(),
      version: '1.0.0',
      agents,
      stats,
    };

    // Ensure output directory exists
    await mkdir(path.dirname(OUTPUT_PATH), { recursive: true });

    // Write registry with pretty formatting
    await writeFile(OUTPUT_PATH, JSON.stringify(registry, null, 2) + '\n');

    console.log('✓ Registry generated successfully!\n');
    console.log('📊 Statistics:');
    console.log(`  • Total agents: ${stats.total}`);
    console.log(`  • By tier:`);
    Object.entries(stats.by_tier)
      .sort(([a], [b]) => a.localeCompare(b))
      .forEach(([tier, count]) => {
        console.log(`    - ${tier}: ${count}`);
      });
    console.log(`  • By category:`);
    Object.entries(stats.by_category)
      .sort(([, a], [, b]) => b - a)
      .forEach(([category, count]) => {
        console.log(`    - ${category}: ${count}`);
      });
    console.log(`  • By complexity:`);
    Object.entries(stats.by_complexity)
      .sort(([, a], [, b]) => b - a)
      .forEach(([complexity, count]) => {
        console.log(`    - ${complexity}: ${count}`);
      });
    console.log(`  • By model:`);
    Object.entries(stats.by_model)
      .sort(([, a], [, b]) => b - a)
      .forEach(([model, count]) => {
        console.log(`    - ${model}: ${count}`);
      });
    console.log(`\n📄 Output: ${OUTPUT_PATH}`);
  } catch (error) {
    console.error('✗ Error generating registry:', error);
    process.exit(1);
  }
}

// ============================================================================
// Entry Point
// ============================================================================

generateRegistry();

File: scripts/generate-registry.py
#!/usr/bin/env python3
"""
Agent Registry Generator

Generates comprehensive agent registry files for the Claude Agents Pro marketplace.

Outputs:
  - agent-metadata.json: Detailed metadata with quality scores and relationships
  - marketplace.json: Public-facing marketplace catalog
  - registry.json: Simplified agent listing

Usage:
    python3 scripts/generate-registry.py
    python3 scripts/generate-registry.py --agents-dir agents --output-metadata configs/agent-metadata.json
"""

import argparse
import json
import re
import sys
from collections import defaultdict
from datetime import datetime
from pathlib import Path
from typing import Any

try:
    import yaml
except ImportError:
    print("Error: PyYAML required. Install with: pip install pyyaml")
    sys.exit(1)


# ============================================================================
# Tier and Category Mappings
# ============================================================================

TIER_CATEGORIES = {
    "00-meta": "Meta-Orchestration",
    "01-foundation": "Foundation",
    "02-development": "Development",
    "03-specialists": "Specialists",
    "04-experts": "Experts",
    "05-domains": "Domains",
    "06-integration": "Integration",
    "07-quality": "Quality",
    "08-automation": "Automation",
    "08-finance": "Finance",
    "09-enterprise": "Enterprise",
}

TIER_DESCRIPTIONS = {
    "00-meta": "Multi-agent workflow orchestration and coordination",
    "01-foundation": "Core engineering agents for essential development tasks",
    "02-development": "Language and platform specialists",
    "03-specialists": "Domain experts for infrastructure and architecture",
    "04-experts": "Advanced specialists for ML, AI, and specialized domains",
    "05-domains": "Domain-specific experts for specialized business contexts",
    "06-integration": "Research, documentation, and knowledge integration",
    "07-quality": "Quality assurance, testing, and validation specialists",
    "08-automation": "Automation and infrastructure tooling experts",
    "08-finance": "Financial analysis and quantitative modeling specialists",
    "09-enterprise": "Enterprise architecture and governance specialists",
}

TIER_COMPLEXITY = {
    "00-meta": "expert",
    "01-foundation": "complex",
    "02-development": "complex",
    "03-specialists": "complex",
    "04-experts": "expert",
    "05-domains": "moderate",
    "06-integration": "moderate",
    "07-quality": "complex",
    "08-automation": "expert",
    "08-finance": "expert",
    "09-enterprise": "expert",
}

# Keywords for capability extraction
CAPABILITY_KEYWORDS = [
    'REST', 'GraphQL', 'gRPC', 'API', 'OpenAPI', 'OAuth', 'JWT',
    'React', 'Vue', 'Angular', 'TypeScript', 'JavaScript', 'Python', 'Go', 'Java',
    'AWS', 'Azure', 'GCP', 'Kubernetes', 'Docker', 'Terraform',
    'PostgreSQL', 'MongoDB', 'Redis', 'DynamoDB',
    'CI/CD', 'DevOps', 'MLOps', 'SRE', 'Security',
    'Machine Learning', 'Deep Learning', 'NLP', 'Computer Vision',
    'Testing', 'Performance', 'Observability', 'Monitoring',
    'Frontend', 'Backend', 'Full Stack', 'Mobile',
    'Microservices', 'Event-Driven', 'CQRS', 'DDD',
]


# ============================================================================
# Utility Functions
# ============================================================================

def extract_frontmatter(content: str) -> dict | None:
    """Extract YAML frontmatter from markdown content."""
    pattern = r"^---\s*\n(.*?)\n---\s*\n"
    match = re.match(pattern, content, re.DOTALL)
    if match:
        try:
            return yaml.safe_load(match.group(1))
        except yaml.YAMLError as e:
            print(f"  YAML parse error: {e}")
            return None
    return None


def extract_tier(file_path: Path, agents_dir: Path) -> str:
    """Extract tier from file path (e.g., '01-foundation')."""
    try:
        relative_path = file_path.relative_to(agents_dir)
        tier_part = relative_path.parts[0] if relative_path.parts else None
        if tier_part and re.match(r"^\d{2}-", tier_part):
            return tier_part
    except ValueError:
        pass
    return "uncategorized"


def normalize_model(model: str | None) -> str:
    """Normalize model name to standard format."""
    if not model:
        return "sonnet"

    model_lower = model.lower()
    if "opus" in model_lower:
        return "opus"
    if "sonnet" in model_lower:
        return "sonnet"
    if "haiku" in model_lower:
        return "haiku"

    return "sonnet"


def extract_capabilities(description: str) -> list[str]:
    """Extract capabilities from description."""
    capabilities = []
    for keyword in CAPABILITY_KEYWORDS:
        if re.search(rf'\b{re.escape(keyword)}\b', description, re.IGNORECASE):
            capabilities.append(keyword)

    return list(set(capabilities))[:10]  # Limit to 10 unique


def extract_tags(description: str, name: str) -> list[str]:
    """Extract tags from description and name."""
    tags = set()

    # Add name-based tags
    name_parts = name.split('-')
    for part in name_parts:
        if len(part) > 3:
            tags.add(part.lower())

    # Extract from description
    tag_keywords = [
        'api', 'rest', 'graphql', 'design', 'architecture', 'security',
        'testing', 'performance', 'optimization', 'monitoring', 'observability',
        'cloud', 'aws', 'kubernetes', 'docker', 'devops', 'ci/cd',
        'frontend', 'backend', 'mobile', 'react', 'typescript', 'python',
        'database', 'sql', 'nosql', 'machine learning', 'ml', 'ai',
        'documentation', 'research', 'review', 'debugging',
    ]

    for keyword in tag_keywords:
        if re.search(rf'\b{re.escape(keyword)}\b', description, re.IGNORECASE):
            tags.add(keyword.lower())

    return sorted(list(tags))[:15]  # Limit to 15 tags


def determine_complexity(description: str, tools: list[str]) -> str:
    """Determine complexity based on description and tools."""
    desc_length = len(description)
    tool_count = len(tools)

    if desc_length > 300 or tool_count >= 6:
        return "complex"
    if desc_length > 150 or tool_count >= 4:
        return "moderate"
    return "simple"


# ============================================================================
# Agent Parsing
# ============================================================================

def parse_agent_file(file_path: Path, agents_dir: Path, project_root: Path) -> dict | None:
    """Parse a single agent markdown file and extract metadata."""
    try:
        content = file_path.read_text(encoding='utf-8')
        frontmatter = extract_frontmatter(content)

        if not frontmatter:
            print(f"⚠ Skipping {file_path.name}: no valid frontmatter")
            return None

        # Validate required fields
        if 'name' not in frontmatter or 'description' not in frontmatter:
            print(f"⚠ Skipping {file_path.name}: missing required fields")
            return None

        name = frontmatter['name']
        description = frontmatter['description']
        tier = extract_tier(file_path, agents_dir)

        # Parse tools
        tools = frontmatter.get('tools', [])
        if isinstance(tools, str):
            tools = [t.strip() for t in re.split(r'[,\s]+', tools) if t.strip()]
        elif not isinstance(tools, list):
            tools = []

        # Extract metadata
        relative_path = file_path.relative_to(project_root)
        model = normalize_model(frontmatter.get('model'))
        complexity = determine_complexity(description, tools)
        capabilities = extract_capabilities(description)
        tags = extract_tags(description, name)

        return {
            "name": name,
            "description": description.strip(),
            "tier": tier,
            "category": TIER_CATEGORIES.get(tier, "Uncategorized"),
            "complexity": complexity,
            "model": model,
            "model_rationale": frontmatter.get('model_rationale', ''),
            "tools": tools,
            "capabilities": capabilities,
            "tags": tags,
            "file": str(relative_path),
            "subagent_type": frontmatter.get('subagent_type', name),
            "enhanced_capabilities": frontmatter.get('enhanced_capabilities', []),
        }

    except Exception as e:
        print(f"✗ Error parsing {file_path.name}: {e}")
        return None


def find_agent_files(agents_dir: Path) -> list[Path]:
    """Find all agent markdown files (excluding README, TESTING, etc.)."""
    excluded = {"README.md", "TESTING.md", "AGENT_CHECKLIST.md"}
    agents = []

    for tier_dir in sorted(agents_dir.iterdir()):
        if tier_dir.is_dir() and re.match(r'^\d{2}-', tier_dir.name):
            for md_file in tier_dir.glob("*.md"):
                # Exclude documentation files and glossaries
                if md_file.name not in excluded and not md_file.name.endswith("-glossary.md"):
                    agents.append(md_file)

    return sorted(agents)


# ============================================================================
# Registry Generation
# ============================================================================

def generate_agent_metadata(agents: list[dict]) -> dict:
    """Generate agent-metadata.json structure."""
    metadata = {
        "schema_version": "1.0",
        "metadata": {}
    }

    # Group by tier
    by_tier = defaultdict(list)
    for agent in agents:
        by_tier[agent['tier']].append(agent)

    # Build metadata structure
    for tier, tier_agents in sorted(by_tier.items()):
        tier_num = int(tier.split('-')[0])
        metadata["metadata"][tier] = {
            "tier": tier_num,
            "category": TIER_CATEGORIES.get(tier, "Uncategorized"),
            "agents": {}
        }

        for agent in tier_agents:
            metadata["metadata"][tier]["agents"][agent['name']] = {
                "quality_score": 8.0,  # Default, can be enhanced with actual scoring
                "dependencies": [],
                "delegates_to": [],
                "tools": agent['tools'],
                "activation_patterns": agent['tags'][:5],
                "last_updated": datetime.now().strftime("%Y-%m-%d"),
                "maintainer": "ubehera"
            }

    return metadata


def generate_marketplace(agents: list[dict]) -> dict:
    """Generate marketplace.json structure."""
    # Group by tier
    by_tier = defaultdict(list)
    for agent in agents:
        by_tier[agent['tier']].append(agent)

    # Build tier information
    tiers = []
    for tier in sorted(by_tier.keys()):
        tier_num = int(tier.split('-')[0])
        tiers.append({
            "id": tier,
            "name": TIER_CATEGORIES.get(tier, "Uncategorized"),
            "description": TIER_DESCRIPTIONS.get(tier, ""),
            "level": tier_num,
            "complexity": TIER_COMPLEXITY.get(tier, "moderate")
        })

    # Build agent entries
    agent_entries = []
    for agent in agents:
        agent_entries.append({
            "id": agent['name'],
            "name": agent['name'],
            "tier": agent['tier'],
            "category": agent['category'],
            "description": agent['description'],
            "model": agent['model'],
            "model_rationale": agent['model_rationale'],
            "complexity": agent['complexity'],
            "capabilities": agent['capabilities'],
            "tags": agent['tags'],
            "tools": agent['tools'],
            "path": agent['file']
        })

    return {
        "version": "1.0.0",
        "last_updated": datetime.now().strftime("%Y-%m-%d"),
        "registry": {
            "name": "Claude Agents Pro Marketplace",
            "description": "Production-ready Claude Code agents with tiered orchestration system",
            "repository": "https://github.com/ubehera/claude-agents-pro",
            "license": "Apache-2.0"
        },
        "tiers": tiers,
        "agents": agent_entries,
        "statistics": {
            "total_agents": len(agents),
            "by_tier": {tier: len(tier_agents) for tier, tier_agents in by_tier.items()},
            "by_model": dict(defaultdict(int, {agent['model']: sum(1 for a in agents if a['model'] == agent['model']) for agent in agents})),
            "by_complexity": dict(defaultdict(int, {agent['complexity']: sum(1 for a in agents if a['complexity'] == agent['complexity']) for agent in agents}))
        }
    }


def generate_registry(agents: list[dict]) -> dict:
    """Generate simplified registry.json structure."""
    # Group by tier
    by_tier = defaultdict(list)
    for agent in agents:
        by_tier[agent['tier']].append(agent)

    agent_entries = []
    for agent in agents:
        agent_entries.append({
            "name": agent['name'],
            "description": agent['description'],
            "tier": agent['tier'],
            "path": agent['file'],
            "tools": agent['tools'],
            "enhanced_capabilities": agent['enhanced_capabilities'],
            "subagent_type": agent['subagent_type']
        })

    return {
        "version": "1.0.0",
        "generated": datetime.now().isoformat(),
        "total_agents": len(agents),
        "agents": agent_entries,
        "tiers": sorted(by_tier.keys())
    }


def calculate_statistics(agents: list[dict]) -> dict:
    """Calculate registry statistics."""
    by_tier = defaultdict(int)
    by_category = defaultdict(int)
    by_complexity = defaultdict(int)
    by_model = defaultdict(int)

    for agent in agents:
        by_tier[agent['tier']] += 1
        by_category[agent['category']] += 1
        by_complexity[agent['complexity']] += 1
        by_model[agent['model']] += 1

    return {
        "total": len(agents),
        "by_tier": dict(by_tier),
        "by_category": dict(by_category),
        "by_complexity": dict(by_complexity),
        "by_model": dict(by_model)
    }


# ============================================================================
# Main
# ============================================================================

def main():
    parser = argparse.ArgumentParser(description="Generate agent registry files")
    parser.add_argument("--agents-dir", type=Path, default=Path("agents"), help="Agents directory")
    parser.add_argument("--output-metadata", type=Path, default=Path("configs/agent-metadata.json"), help="Agent metadata output")
    parser.add_argument("--output-marketplace", type=Path, default=Path("configs/marketplace.json"), help="Marketplace output")
    parser.add_argument("--output-registry", type=Path, default=Path("registry.json"), help="Registry output")
    parser.add_argument("--quiet", action="store_true", help="Suppress output")
    args = parser.parse_args()

    # Find project root
    script_dir = Path(__file__).parent
    project_root = script_dir.parent

    agents_dir = project_root / args.agents_dir
    if not agents_dir.exists():
        print(f"Error: Agents directory not found at {agents_dir}")
        sys.exit(1)

    # Find and parse agents
    if not args.quiet:
        print("🚀 Generating agent registry...\n")

    agent_files = find_agent_files(agents_dir)
    if not agent_files:
        print("Error: No agent files found")
        sys.exit(1)

    if not args.quiet:
        print(f"Found {len(agent_files)} agent files to process...")

    agents = []
    for file_path in agent_files:
        agent = parse_agent_file(file_path, agents_dir, project_root)
        if agent:
            agents.append(agent)

    if not agents:
        print("Error: No valid agents parsed")
        sys.exit(1)

    # Sort by tier and name
    agents.sort(key=lambda a: (a['tier'], a['name']))

    # Generate outputs
    metadata = generate_agent_metadata(agents)
    marketplace = generate_marketplace(agents)
    registry = generate_registry(agents)
    stats = calculate_statistics(agents)

    # Write outputs
    output_files = [
        (args.output_metadata, metadata),
        (args.output_marketplace, marketplace),
        (args.output_registry, registry),
    ]

    for output_path, data in output_files:
        full_path = project_root / output_path
        full_path.parent.mkdir(parents=True, exist_ok=True)

        with open(full_path, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2, ensure_ascii=False)
            f.write('\n')

        if not args.quiet:
            print(f"✓ Generated: {full_path.relative_to(project_root)}")

    # Print statistics
    if not args.quiet:
        print("\n📊 Statistics:")
        print(f"  • Total agents: {stats['total']}")
        print("  • By tier:")
        for tier, count in sorted(stats['by_tier'].items()):
            print(f"    - {tier}: {count}")
        print("  • By category:")
        for category, count in sorted(stats['by_category'].items(), key=lambda x: -x[1]):
            print(f"    - {category}: {count}")
        print("  • By complexity:")
        for complexity, count in sorted(stats['by_complexity'].items(), key=lambda x: -x[1]):
            print(f"    - {complexity}: {count}")
        print("  • By model:")
        for model, count in sorted(stats['by_model'].items(), key=lambda x: -x[1]):
            print(f"    - {model}: {count}")

    print("\n✅ Registry generation complete!")


if __name__ == "__main__":
    main()
